File: backend/app/schemas/role.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import PaginationParams, build_pages
from app.schemas.permission import PermissionRead


class RoleCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    name: str = Field(min_length=1, max_length=50)
    description: str | None = Field(default=None, max_length=255)


class RoleUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    name: str | None = Field(default=None, min_length=1, max_length=50)
    description: str | None = Field(default=None, max_length=255)


class RoleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    created_at: datetime
    updated_at: datetime


class RoleWithPermissionsRead(RoleRead):
    permissions: list[PermissionRead]


class RoleListResponse(BaseModel):
    items: list[RoleRead]
    total: int
    page: int
    size: int
    pages: int

    @classmethod
    def build(
        cls,
        items: list[RoleRead],
        total: int,
        pagination: PaginationParams,
    ) -> "RoleListResponse":
        return cls(
            items=items,
            total=total,
            page=pagination.page,
            size=pagination.size,
            pages=build_pages(total, pagination.size),
        )

File: frontend/src/types/role.ts
import type { Permission } from './permission'

export interface Role {
  id: number
  name: string
  description: string | null
  created_at: string
  updated_at: string
}

export interface RoleWithPermissions extends Role {
  permissions: Permission[]
}

export interface RoleCreate {
  name: string
  description?: string | null
}

export interface RoleUpdate {
  name?: string
  description?: string | null
}

export interface RoleListResponse {
  items: Role[]
  total: number
  page: number
  size: number
  pages: number
}
